module/zfs/zfs_vnops.c

```diff
@@ -2950,25 +2950,27 @@ zfs_setattr(struct inode *ip, vattr_t *vap, int flags, cred_t *cr)
 	if (mask & (ATTR_UID|ATTR_GID)) {
 
 		if (mask & ATTR_UID) {
-			SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_UID(zsb), NULL,
-			    &new_uid, sizeof (new_uid));
 			ZTOI(zp)->i_uid = SUID_TO_KUID(new_uid);
+			new_kuid = zfs_uid_read(ZTOI(zp));
+			SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_UID(zsb), NULL,
+			    &new_kuid, sizeof (new_kuid));
```

Comment:
Using the name kuid here can be confusing since kuid_t is what inode->i_uid is using. You can use something like sauid. Also, consider putting the declaration inside the if-scope, since it's very short living.